components/object_store/src/obkv/mod.rs

```diff
@@ -0,0 +1,940 @@
+        debug!(
+            "ObkvObjectStore list_with_delimiter operation, prefix:{path}, cost:{}",
+            instant.elapsed().as_millis()
+        );
+        Ok(ListResult {
+            common_prefixes,
+            objects,
+        })
+    }
+
+    async fn copy(&self, _from: &Path, _to: &Path) -> Result<()> {
+        // TODO:
+        Err(StoreError::NotImplemented)
+    }
+
+    async fn copy_if_not_exists(&self, _from: &Path, _to: &Path) -> Result<()> {
+        // TODO:
+        Err(StoreError::NotImplemented)
+    }
+}
+
+struct ObkvMultiPartUpload<T> {
+    /// The full path to the object.
+    location: Path,
+    /// The id of multi upload tasks, we use this id as object version.
+    upload_id: String,
+    /// The table name of obkv to save data.
+    table_name: String,
+    /// The client of object store.
+    client: Arc<T>,
+    /// The size of object.
+    size: AtomicU64,
+    /// The size in bytes of one part. Note: maybe the size of last part less
+    /// than part_size.
+    part_size: usize,
+    /// The mananger to process meta info.
+    meta_manager: Arc<MetaManager<T>>,
+}
+
+struct Prefix;
+
+impl Prefix {
+    fn build(location: &Path, upload_id: &str) -> String {
+        format!("{location}@{upload_id}@")
+    }
+}
+
+#[async_trait]
+impl<T: TableKv> CloudMultiPartUploadImpl for ObkvMultiPartUpload<T> {
+    async fn put_multipart_part(
+        &self,
+        buf: Vec<u8>,
+        part_idx: usize,
+    ) -> Result<UploadPart, std::io::Error> {
+        let mut batch = T::WriteBatch::default();
+        let prefix = Prefix::build(&self.location, &self.upload_id);
+        let key = format!("{prefix}{part_idx}");
+        batch.insert(key.as_bytes(), buf.as_ref());
+
+        self.client
+            .write(WriteContext::default(), &self.table_name, batch)
+            .map_err(|source| StoreError::Generic {
+                store: OBKV,
+                source: Box::new(source),
+            })?;
+        // Record size of object.
+        self.size.fetch_add(buf.len() as u64, Ordering::Relaxed);
+        Ok(UploadPart { content_id: key })
+    }
+
+    async fn complete(&self, completed_parts: Vec<UploadPart>) -> Result<(), std::io::Error> {
+        // We should save meta info after finish save data.
+        let mut paths = Vec::with_capacity(completed_parts.len());
+        for upload_part in completed_parts {
+            paths.push(upload_part.content_id);
+        }
+
+        let now = SystemTime::now();
+        let since_epoch = now.duration_since(UNIX_EPOCH).expect("Time went backwards");
+        let last_modified = since_epoch.as_millis() as i64;
+
+        let meta = ObkvObjectMeta {
+            location: self.location.as_ref().to_string(),
+            last_modified,
+            size: self.size.load(Ordering::SeqCst) as usize,
+            unique_id: Some(format!(
+                "{}@{}@{}",
+                self.table_name, self.location, self.upload_id
+            )),
```

Comment:
Is this formatting the same as the `Prefix::build`? Maybe we should put them together.